network/src/sync/blocks/aggro.rs

```diff
@@ -0,0 +1,201 @@
+                .map(|peer| (peer.clone(), block_locator_hashes.hashes.clone()))
+                .collect()
+        };
+
+        let hash_requests_sent = self.send_sync_messages(peer_syncs).await?;
+
+        if hash_requests_sent == 0 {
+            return Ok(());
+        }
+
+        let received_hashes = Arc::new(RwLock::new(Cache::<1024, 32>::default()));
+
+        let node = self.base.node.clone();
+        self.base
+            .receive_messages(15, 3, |msg| {
+                metrics::decrement_gauge!(snarkos_metrics::queues::SYNC_ITEMS, 1.0);
+                match msg {
+                    SyncInbound::BlockHashes(peer, hashes) => {
+                        debug!("received {} sync hashes from {}", hashes.len(), peer);
+                        let hashes: Vec<Digest> = hashes.into_iter().map(|x| x.0.into()).collect::<Vec<_>>();
+                        if hashes.is_empty() {
+                            return false;
+                        }
+                        let last_hash = hashes.last().unwrap().clone();
+
+                        let node = node.clone();
+                        let block_locator_hashes = block_locator_hashes.clone();
+                        let received_hashes = received_hashes.clone();
+                        tokio::spawn(async move {
+                            if block_locator_hashes.read().await.last_update.elapsed() > Duration::from_secs(10) {
+                                match SyncBase::block_locator_hashes(&node).await {
+                                    Ok(hashes) => {
+                                        let mut target = block_locator_hashes.write().await;
+                                        target.hashes = hashes;
+                                        target.last_update = Instant::now();
+                                    }
+                                    Err(e) => warn!("sync failed to fetch block locator hashes: {:?}", e),
+                                }
+                            }
+
+                            let mut hashes_trimmed = Vec::with_capacity(hashes.len());
+                            {
+                                let received_hashes = received_hashes.read().await;
+                                for hash in hashes {
+                                    if !received_hashes.contains(&hash[..]) {
+                                        hashes_trimmed.push(hash);
+                                    }
+                                }
+                            }
+
+                            let early_block_states = match node.storage.get_block_states(&hashes_trimmed[..]).await {
+                                Ok(x) => x,
+                                Err(e) => {
+                                    warn!("failed to get block states: {:?}", e);
+                                    return;
+                                }
+                            };
+
+                            let blocks: IndexSet<_> = {
```

Comment:
do we need an `IndexSet` (as opposed to a plain `HashSet`) here?

Reply:
no -- but i appreciate the added determinism as a hedge against non-determinism bugs -- and as usual, not perf critical